collector/src/main/java/io/prometheus/jmx/JmxCollector.java

```diff
@@ -312,29 +323,22 @@ public void recordBean(
         // attrDescription tends not to be useful, so give the fully qualified name too.
         String help = attrDescription + " (" + beanName + attrName + ")";
 
-        String attrNameSnakeCase = snakeCasePattern.matcher(attrName).replaceAll("$1_$2").toLowerCase();
+        String attrNameSnakeCase = toSnakeCase(attrName);
+
+        // Evict patterns no longer listed in the rules
+        // TODO
 
         for (Rule rule : config.rules) {
           Matcher matcher = null;
           String matchName = beanName + (rule.attrNameSnakeCase ? attrNameSnakeCase : attrName);
           if (rule.pattern != null) {
-            matcher = rule.pattern.matcher(matchName + ": " + beanValue);
-            if (!matcher.matches()) {
-              continue;
+            if (!rulePatternCache.checkAndStoreMatchName(rule.pattern, matchName + ": ")) {
```

Comment:
`beanValue` is no longer a part of the regex processing. If a user has `beanValue` in their regex pattern, it'll be ignored. None of the example yamls do this, so I'm wondering what the utility of `beanValue` is for users.

Reply:
Unfortunately, including `beanValue` in the regex causes the cache not to work because the value is more than likely different every scrape.

Reply:
This is supported and used behaviour, it cannot be broken.

Reply:
Is the purpose of the feature to have metrics show up in users' metric names? 

I wonder if I can check if the the supplied `pattern` contains `: ` and if yes, don't use the cache.

Reply:
Some mBeans contain strings as values and this allows working with them. There's no real way to check for that.

Reply:
I'll keep thinking about what to do here. In the meantime, I've rebased https://github.com/prometheus/jmx_exporter/pull/235 and split off the snakeCase change from this PR into this one: https://github.com/prometheus/jmx_exporter/pull/236

Reply:
What if I add an opt-in yaml param parallel to pattern with default == true: enableBeanValueInPattern

For for example kafka.yaml would look like this
```
lowercaseOutputName: true
enableBeanValueInPattern: false
rules:
- pattern : kafka.cluster<type=(.+), name=(.+), topic=(.+), partition=(.+)><>Value
  name: kafka_cluster_$1_$2
  labels:
    topic: "$3"
    partition: "$4"
...
```

Reply:
That's still a pretty big breaking change, as users will presume the pattern looks the old way.